### PR TITLE
`Development`: Adapt LTI advantage deep linking service for exercise selection from Moodle

### PR DESCRIPTION
**Pull request by basak-akan**
> [!WARNING]  
> **Deploy only locally or on ts11!** The LTI feature is only available on TS11.

#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).
#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added pre-authorization annotations according to the [guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
- [x] I documented the Java code using JavaDoc style.
#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context

Deep Linking in the context of LTI Advantage refers to a standard that allows instructors to create links directly to specific course materials or exercises within a learning platform like Artemis. It makes it easier to integrate external learning tools and content into a Learning Management System like Moodle. 
The deep linking service simplifies the workflow by eliminating manual copy/paste exercise links to Moodle. 

### Description

This update introduces a new deep linking service compliant with the [LTI Deep Linking Specifications](https://www.imsglobal.org/spec/lti-dl/v2p0). It simplifies the instructors' workflow in Moodle by enabling them to embed direct links to Artemis exercises, eliminating the need for complex navigation or manual link copying. Additionally, a user-friendly exercise selection interface is now accessible through Moodle via a pop-up frame after clicking on 'Select Content.' Instructors can choose the desired exercise, and upon clicking 'Link,' the exercise URL is automatically copied to Moodle inside a JWT through an auto-submit form (implemented according to specification). Future enhancements include a new course selection page and the ability to link multiple exercises in a single operation, after implementation of #7185.

### Steps for Testing

Prerequisites:
- Please use TS11 to test this PR since lti profile is enabled
- 1 Instructor

1. Log in to Artemis with an instructor credentials ( e.g. artemis_test_user_11)
2. Navigate to https://moodle.ase.in.tum.de/
3. Login using the same credentials as Artemis
4. Navigate to "LTI Testing Course"
5. Enable "Edit Mode" in the top right corner.
6. Click on "Add activity or resource" under any topic.
7. A pop-up window will come out. Select "External Tool"
8. You will be navigated to the "Adding a new External Tool" page
9. On this page, select preconfigured tool as "TS11 - ltitestcourse"
10.  If you may encounter "Required Activity Name" warning please ignore it and follow the remaining steps.
11.  After selecting the preconfigured tool click on the "Select Content" button
12. An Artemis frame will pop up. Verify this frame contains exercises from "LTI Testing Course" 
13.  Select one of the exercises and click the "Link" button 
14. Verify the exercise link appeared on the "Tool Url" input field.
15. Select save and display at the bottom of the page.
16. Verify you can see the Artemis exercise.

#### Performance Review
- [x] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [x] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

#### Client

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [course-lti-configuration.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5605/latest/artifact/shared/Coverage-Report-Client-Tests/app/course/manage/course-lti-configuration/course-lti-configuration.component.ts.html) | 100% | ✅ |
| [lti13-deep-linking.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5605/latest/artifact/shared/Coverage-Report-Client-Tests/app/lti/lti13-deep-linking.component.ts.html) | 89.83% | ✅ |
| [lti13-exercise-launch.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5605/latest/artifact/shared/Coverage-Report-Client-Tests/app/lti/lti13-exercise-launch.component.ts.html) | 95.52% | ✅ |
| [lti13-select-content.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5605/latest/artifact/shared/Coverage-Report-Client-Tests/app/lti/lti13-select-content.component.ts.html) | 91.66% | ✅ |
| [lti-initializer-modal.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5605/latest/artifact/shared/Coverage-Report-Client-Tests/app/overview/exercise-details/lti-initializer-modal.component.ts.html) | 78.57% | ✅ |
| [lti-initializer.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5605/latest/artifact/shared/Coverage-Report-Client-Tests/app/overview/exercise-details/lti-initializer.component.ts.html) | 100% | ✅ |

#### Server

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [CustomLti13Configurer.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5605/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.config.lti/CustomLti13Configurer.html) | 100% | ✅ |
| [Claims.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5605/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.domain.lti/Claims.html) | 100% | ✅ |
| [Lti13ClientRegistration.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5605/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.domain.lti/Lti13ClientRegistration.html) | 83% | ✅ |
| [Lti13DeepLinkingResponse.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5605/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.domain.lti/Lti13DeepLinkingResponse.html) | 82% | ✅ |
| [LtiAuthenticationResponseDTO.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5605/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.domain.lti/LtiAuthenticationResponseDTO.html) | 100% | ✅ |
| [Lti13LaunchFilter.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5605/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.security.lti/Lti13LaunchFilter.html) | 100% | ✅ |
| [Lti13TokenRetriever.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5605/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.security.lti/Lti13TokenRetriever.html) | 98% | ✅ |
| [Lti13Service.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5605/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.connectors.lti/Lti13Service.html) | 93% | ✅ |
| [LtiDeepLinkingService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5605/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.connectors.lti/LtiDeepLinkingService.html) | 96% | ✅ |
| [LtiResource.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5605/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest/LtiResource.html) | 69% | ✅ |

### Screenshots

deep_linking

### CODE REVIEW

**Review by JanaNF — APPROVED** [review]
Manual tested on legacy ts11. Worked as expected.

Note: As in your video, when adding a new external tool, I got the "Required Activity Name" warning even though I filled it out. I was still able to add the tool, but I was a little confused at first. If you could change this, it would be very helpful.

**Review by farisd16 — APPROVED** [review]
tested on legacy ts11, works as expected

**Review by Gusti2010 — DISMISSED** [review]
Tested manually on TS11. Works as expected!

**Review by b-fein — CHANGES_REQUESTED** [review]
Code looks mostly good.

I don’t thing the `LtiDeepLinkingService` works as intended, though, in case there are multiple deep linking requests at the same time.

In client code a few constructs could be simplified by using Angular functionality.

**Comment by basak-akan** [comment]
> Code looks mostly good.
> 
> I don’t thing the `LtiDeepLinkingService` works as intended, though, in case there are multiple deep linking requests at the same time.
> 
> In client code a few constructs could be simplified by using Angular functionality.

@b-fein Thank you for your insights on the code review. I've taken your observations into account and refactored the LtiDeepLinkingService to handle concurrent deep-linking requests. Also, I've simplified the client code constructs by integrating more of Angular's built-in features as suggested. Could you please review the changes again?

**Review by JanaNF — COMMENTED** [review]
Tried to test on legacy_ts11 but I couldn’t connect to the server.

I wrote with Laurenz and he said:
_"A new config variable was introduced by a PR, but this is not included in the config for TS11. This is why it always fails at startup. I think the PR was merged without regard to the config change and is now causing problems on all test servers where the config has not yet been adjusted. The PR with the config change is already in Develop and I suspect that if your PRs are running with the latest changes from Develop, then there will be problems"_

I think your PR is affected.

**Review by JanaNF — APPROVED** [review]
I could your PR on legacy_ts11 again :tada: Everything still works

File: src/test/java/de/tum/in/www1/artemis/connectors/LtiDeepLinkingServiceTest.java [hunk 1]
```diff
@@ -0,0 +1,150 @@
+package de.tum.in.www1.artemis.connectors;
```
Comment (b-fein) [inline src/test/java/de/tum/in/www1/artemis/connectors/LtiDeepLinkingServiceTest.java:1]:
```suggestion
package de.tum.in.www1.artemis.connectors.lti;
```
If the package matches the production package structure methods only used in tests could be made package-private.

**Review by b-fein — APPROVED** [review]
Thanks for implementing the previous suggestions. Code looks good with only a few tiny questions and optional suggestions.

**Review by AntonGeTUM — DISMISSED** [review]
Tested on TS11, works as expected

File: src/main/webapp/app/lti/lti13-deep-linking.component.html [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+            <jhi-help-icon class="ps-1" text="artemisApp.lti13.selectContentTooltip"></jhi-help-icon>
+        </div>
+    </div>
+    <div class="modal-body">
+        <div class="table-responsive">
+            <table class="table table-bordered">
+                <thead class="thead-dark">
+                    <tr jhiSort [(predicate)]="predicate" [(ascending)]="reverse" (sortChange)="sortRows()">
+                        <th class="d-md-table-cell">
+                            <!-- <input class="form-check-input" type="checkbox" />-->
```
[the remark this answers is not in the document]
Comment (basak-akan) [reply src/main/webapp/app/lti/lti13-deep-linking.component.html]: Somehow I missed to remove it, it is removed now.

**Review by b-fein — APPROVED** [review]
Code looks good now. :+1:

**Review by JanaNF — DISMISSED** [review]
Re-tested on legacy_ts11. Worked as expected.

File: src/test/java/de/tum/in/www1/artemis/LtiIntegrationTest.java [hunk 3]
```diff
@@ -354,4 +392,91 @@ private void assertParametersNewStudent(MultiValueMap<String, String> parameters
...
+        claims.put("name", "Ms Jane Marie Doe");
+        claims.put("given_name", "Jane");
+        claims.put("family_name", "Doe");
+        claims.put("middle_name", "Marie");
+        claims.put("picture", "https://example.org/jane.jpg");
+        claims.put("email", "jane@example.org");
+        claims.put("locale", "en-US");
+
+        // LTI specific claims
+        claims.put("https://purl.imsglobal.org/spec/lti/claim/deployment_id", "07940580-b309-415e-a37c-914d387c1150");
```
Comment (lennart-keller) [inline src/test/java/de/tum/in/www1/artemis/LtiIntegrationTest.java:431]: Maybe you can reuse the constants for the URLs here :)

**Review by lennart-keller — COMMENTED** [review]
Code looks good to me. Left a few questions

File: src/main/java/de/tum/in/www1/artemis/domain/lti/Lti13DeepLinkingResponse.java [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParser;
+
+/**
+ * Represents the LTI 1.3 Deep Linking Response.
+ * It encapsulates the necessary information to construct a valid deep linking response
+ * according to the LTI 1.3 specification.
+ */
+public class Lti13DeepLinkingResponse {
```
[the remark this answers is not in the document]
Comment (basak-akan) [reply src/main/java/de/tum/in/www1/artemis/domain/lti/Lti13DeepLinkingResponse.java:17]: I am not sure if it fits the criteria for being a record. It contains a custom constructor and methods (validateClaims, getClaims). But sure, I can transfer these methods to LtiDeepLinkingService and implement a static factory method to replace the custom constructor. What do you think?

[the remark this answers is not in the document]
Comment (lennart-keller) [reply src/main/java/de/tum/in/www1/artemis/domain/lti/Lti13DeepLinkingResponse.java:17]: I guess it's fine to keep it like this.

**Review by lennart-keller — APPROVED** [review]
code lgtm. Thanks for addressing my feedback

**Review by laurenzfb — DISMISSED** [review]
Tested in testing session. Worked as expected!

**Review by sarpsahinalp — DISMISSED** [review]
Tested in testing session. Worked as expected!

**Review by b-fein — DISMISSED** [review]
Code re-approve

**Review by vinceclifford — DISMISSED** [review]
Works as expected, tested during testing session

**Review by laurenzfb — APPROVED** [review]
Reapprove

**Review by b-fein — APPROVED** [review]

**Review by lennart-keller — APPROVED** [review]
Recent code changes lgtm

**Review by vinceclifford — APPROVED** [review]
Reapproval

**Review by sarpsahinalp — APPROVED** [review]
Re-approve

**Pull request merged**